### ISSUE TITLE
Data Structure Writeup

### ISSUE THREAD

**Issue by alaraipek** [alaraipek/Issues#15]
**Collections**
[Updated Blog
](https://alaraipek.github.io/Alara2.0/2024/04/24/CPT-Writeup_IPYNB_2_.html)
Blog Python Model code and SQLite Database.

From VSCode using SQLite3 Editor, show your unique collection/table in database, display rows and columns in the table of the SQLite database.
- Your unique collection or table in the SQLite database represents the structure where your data will be stored. You would typically define this table using SQL commands. For instance, if you were creating a table to store user information, you might have columns for user_id, username, email, etc.
![image](https://github.com/alaraipek/Issues/assets/115954616/d5988a83-f80e-486a-807c-d11a8dbbd64a)
From VSCode model, show your unique code that was created to initialize table and create test data.
![image](https://github.com/alaraipek/Issues/assets/115954616/94da6711-585a-40c2-add6-84c0865a0db6)

**Lists and Dictionaries**
Blog Python API code and use of List and Dictionaries.
In VSCode using Debugger, show a list as extracted from database as Python objects.
In VSCode use Debugger and list, show two distinct example examples of dictionaries, show Keys/Values using debugger.
![image](https://github.com/alaraipek/Issues/assets/115954616/b3cd2deb-05df-4cbb-91c3-865a39b8aa3e)

Blog Python API Code and Use of Lists and Dictionaries:
In your Python API code, you likely use lists and dictionaries to manage and manipulate data. Lists are ordered collections of items, while dictionaries are unordered collections of key-value pairs.
Debugger in VSCode to Display Lists and Dictionaries:
Using the debugger in VSCode, you can inspect the data retrieved from a database as Python objects. This involves extracting a list of data from the database and examining it in the debugger interface.
Examples of Dictionaries Using Debugger:
The debugger can also be used to display examples of dictionaries. You can show two distinct dictionaries, each with its keys and corresponding values, using the debugger interface.

**APIs and JSON**
Blog Python API code and use of Postman to request and respond with JSON.

In VSCode, show Python API code definition for request and response using GET, POST, UPDATE methods. Discuss algorithmic condition used to direct request to appropriate Python method based on request method.
In VSCode, show algorithmic conditions used to validate data on a POST condition.
In Postman, show URL request and Body requirements for GET, POST, and UPDATE methods.
In Postman, show the JSON response data for 200 success conditions on GET, POST, and UPDATE methods.
![image](https://github.com/alaraipek/Issues/assets/115954616/ae642375-2755-4a69-9cc1-2548a785c5be)
![image](https://github.com/alaraipek/Issues/assets/115954616/2504cc22-fa63-42bc-85e2-80a3c0a66037)
![image](https://github.com/alaraipek/Issues/assets/115954616/0841ba77-f739-45d1-9d0f-1dff974a7c5f)

In Postman, show the JSON response for error for 400 when missing body on a POST request.
In Postman, show the JSON response for error for 404 when providing an unknown user ID to a UPDATE request.
![image](https://github.com/alaraipek/Issues/assets/115954616/fe12276a-5dcf-4498-a445-a92aab17768f)

**Frontend**
Blog JavaScript API fetch code and formatting code to display JSON.
![image](https://github.com/alaraipek/Issues/assets/115954616/723993ec-c642-46b3-9447-67bf52420dc3)
In Chrome inspect, show response of JSON objects from fetch of GET, POST, and UPDATE methods.
![image](https://github.com/alaraipek/Issues/assets/115954616/4a625876-3277-4c8d-9a74-74a66134a882)
In the Chrome browser, show a demo (GET) of obtaining an Array of JSON objects that are formatted into the browsers screen.
![image](https://github.com/alaraipek/Issues/assets/115954616/cb4c35fa-8fe3-48d4-b81e-6c58d4d3ac01)
In JavaScript code, describe fetch and method that obtained the Array of JSON objects.
In JavaScript code, show code that performs iteration and formatting of data into HTML.
![image](https://github.com/alaraipek/Issues/assets/115954616/c93cc4fc-fd7c-4323-93d1-c6fd85a86da6)
In the Chrome browser, show a demo (POST or UPDATE) gathering and sending input and receiving a response that show update. Repeat this demo showing both success and failure.
![image](https://github.com/alaraipek/Issues/assets/115954616/20efe72b-e3f5-401e-bdf3-1d6f90338bfa)

In JavaScript code, show and describe code that handles success. Describe how code shows success to the user in the Chrome Browser screen.
In JavaScript code, show and describe code that handles failure. Describe how the code shows failure to the user in the Chrome Browser screen.
![image](https://github.com/alaraipek/Issues/assets/115954616/ad642109-0d18-4bd9-9b52-4d3c444a82d2)

Optional/Extra, Algorithm Analysis
In the ML projects, there is a great deal of algorithm analysis. Think about preparing data and predictions.
![image](https://github.com/alaraipek/Issues/assets/115954616/c5abfcec-343f-497b-a7cd-ed5c00d3b07f)

1. Show algorithms and preparation of data for analysis. This includes cleaning, encoding, and one-hot encoding.
Data Cleaning:
Algorithm: Use techniques like dropping missing values and removing outliers.
Example: cleaned_data = original_data.dropna()
Encoding Categorical Variables:
Algorithm: Convert categorical variables into numerical format using techniques like one-hot encoding.
Example: encoded_data = pd.get_dummies(data, columns=['categorical_column'])
One-Hot Encoding:
Algorithm: Convert categorical variables into binary columns representing each category.
Example: encoded_data = pd.get_dummies(data, columns=['categorical_column'])

2. Show algorithms and preparation for predictions:
Feature Scaling:
Algorithm: Scale numerical features to a standard range to ensure consistency.
Example:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_features = scaler.fit_transform(features)
Model Training:
Algorithm: Choose an appropriate model and train it on the prepared data.
Example:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)
Making Predictions:
Algorithm: Use the trained model to make predictions on new data.
Example:
predictions = model.predict(X_test)
3. Discuss concepts and understanding of Linear Regression algorithms:
Concept: Linear Regression models the relationship between independent variables (features) and a dependent variable (target) by fitting a linear equation to observed data.
Assumptions: Linearity, Independence, Homoscedasticity.
Evaluation: Metrics like Mean Squared Error, R-squared, and Root Mean Squared Error are commonly used to evaluate Linear Regression models.
4. Discuss concepts and understanding of Decision Tree analysis algorithms:
Concept: Decision Trees recursively split the dataset into subsets based on the most significant feature, creating a tree-like structure of decisions.
Splitting Criteria: Criteria like Gini impurity or Information Gain are used to decide how to split the data at each node.
Advantages: Easy to interpret and visualize, handle both numerical and categorical data, and require little data preprocessing.
Disadvantages: Prone to overfitting and sensitive to small variations in the data.
Ensemble Methods: Techniques like Random Forests and Gradient Boosted Trees are used to improve Decision Tree performance and mitigate overfitting.

**Comment by AnanyaAsudani** [comment]
0.94/1
Completed all the criteria and did extra credit. Good organization, screenshots are clear, and complete all the requirements and shows clear understanding of CRUD. Nice job on the ML and good understanding of data structures.
